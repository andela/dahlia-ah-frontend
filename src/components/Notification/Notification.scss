.notification-container {
  padding: 0 10px;
  display: flex;
  flex-direction: column;
  box-sizing: border-box;
  .notification-desc {
    border: none;
  }
  p {
    padding-left: 10px;
  }

  .not-read {
    background: #f1f1f1;
    width: 100%;
  }

  .mark-all {
    background: white;
    border: none;
  }
  .mark-all:hover, .notification-desc:hover {
    color:  #2d9cdb;
  }
  .notification-header {
    display: flex;
    position: sticky;
    line-height: 0;
    padding: 10px 0;
    top: 0;
    background: white;
    z-index: 1000;
    width: 100%;
    margin-bottom: 5px;
    align-items: baseline;
    justify-content: space-between;
  }
}