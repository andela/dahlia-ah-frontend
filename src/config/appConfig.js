export default {
  FRONTEND_PATH: 'http://localhost:3000',
  BASE_PATH: 'http://localhost:5000',
  BACKEND_PATH: 'http://localhost:5000/api/v1',
};
