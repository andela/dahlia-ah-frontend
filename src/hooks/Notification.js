import { useState } from 'react';

export default () => {
  const [notifications, setNotification] = useState([]);

  return { notifications, setNotification };
};
