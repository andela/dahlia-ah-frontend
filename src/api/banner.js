import axios from 'axios';
import appConfig from '../config/appConfig';

const { BACKEND_PATH } = appConfig;
console.log(BACKEND_PATH);
export default (limit = 1, setBanner) => {
  axios.get(`${BACKEND_PATH}/novels/random?limit=${limit}`)
    .then(({ data }) => {
      setBanner(data.data);
    });
};
