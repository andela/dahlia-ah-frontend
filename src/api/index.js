import axios from 'axios';
import appConfig from '../config/appConfig';

const { BACKEND_PATH } = appConfig;

const user = JSON.parse(window.localStorage.getItem('AuthorsHavenUser'));

if (user) {
  const { token } = user;
  axios.defaults.headers.common.Authorization = token;
}

axios.defaults.baseURL = BACKEND_PATH;

export default axios;
