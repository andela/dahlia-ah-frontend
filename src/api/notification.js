import axios from './index';
import appConfig from '../config/appConfig';

const { BACKEND_PATH } = appConfig;

export const markAllAsRead = (notifications, setNotification) => {
  axios.patch(`${BACKEND_PATH}/notification`).then(() => {
    const updatedNotification = [...notifications];
    updatedNotification.forEach((notification) => {
      notification.isRead = true;
    });
    setNotification(updatedNotification);
  })
    .catch(() => {});
};

export const markAsRead = (index, notifications, setNotification) => {
  const { id, isRead } = notifications[index];
  if (isRead) {
    return;
  }
  axios.patch(`${BACKEND_PATH}/notification/${id}`).then(() => {
    const updatedNotification = [...notifications];
    updatedNotification[index].isRead = true;
    setNotification(updatedNotification);
  })
    .catch(() => {});
};
