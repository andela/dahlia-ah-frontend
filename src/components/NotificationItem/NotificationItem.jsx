import React from 'react';
import moment from 'moment';
import PropTypes from 'prop-types';
import { markAsRead } from '../../api/notification';

const NotificationItem = ({
  notification, index, notifications, setNotification,
}) => {
  const classList = ['collection-item', 'avatar'];
  if (!notification.isRead) {
    classList.push('not-read');
  }
  return (
    <ul
      onClick={() => markAsRead(index, notifications, setNotification)}
      onKeyPress={() => markAsRead(index, notifications, setNotification)}
      className="collection"
      role="presentation"
    >
      <li id="notItem" className={classList.join(' ')}>
        <img id="notImage" src="https://img.icons8.com/color/50/000000/person-male.png" alt={notification.actor} />
        <p id="notText">
          { notification.actor }
          {' '}
          { notification.message }
          {' '}
          { `"${notification.novelTitle}"` }
          <br />
          {' '}
          {moment(notification.createdAt).utc().fromNow()}
        </p>
      </li>
    </ul>
  );
};


NotificationItem.propTypes = {
  notification: PropTypes.shape({
    id: PropTypes.string.isRequired,
    actor: PropTypes.string.isRequired,
    message: PropTypes.string.isRequired,
    novelTitle: PropTypes.string.isRequired,
    novelUrl: PropTypes.string.isRequired,
    createdAt: PropTypes.string.isRequired,
    isRead: PropTypes.bool.isRequired,
  }).isRequired,
  index: PropTypes.number.isRequired,
  notifications: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.string.isRequired,
    actor: PropTypes.string.isRequired,
    message: PropTypes.string.isRequired,
    novelTitle: PropTypes.string.isRequired,
    novelUrl: PropTypes.string.isRequired,
    createdAt: PropTypes.string.isRequired,
    isRead: PropTypes.bool.isRequired,
  })).isRequired,
  setNotification: PropTypes.func.isRequired,
};

export default NotificationItem;
