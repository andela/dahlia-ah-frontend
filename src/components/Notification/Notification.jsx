import React from 'react';
import PropTypes from 'prop-types';
import './Notification.scss';
import NotificationItem from '../NotificationItem/NotificationItem';

const Notification = ({ notifications, setNotification, markAllAsRead }) => (
  <div id="notification" className="notification-container">
    <div className="notification-header">
      <span>Notifications</span>
      <div>
        <button id="mark-all" onClick={() => markAllAsRead(notifications, setNotification)} className="mark-all" type="button">Mark All As Read</button>
      </div>
    </div>
    <div>
      {notifications.length > 0 ? notifications.map((notification, index) => (
        <NotificationItem
          key={notification.id}
          notifications={notifications}
          setNotification={setNotification}
          index={index}
          notification={notification}
        />
      ))
        : <p>You don&#39;t have any notification yet</p>}
    </div>
  </div>
);

Notification.propTypes = {
  notifications: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.string.isRequired,
    actor: PropTypes.string.isRequired,
    message: PropTypes.string.isRequired,
    novelTitle: PropTypes.string.isRequired,
    novelUrl: PropTypes.string.isRequired,
    createdAt: PropTypes.string.isRequired,
    isRead: PropTypes.bool.isRequired,
  })).isRequired,
  markAllAsRead: PropTypes.func.isRequired,
  setNotification: PropTypes.func.isRequired,
};

export default Notification;
