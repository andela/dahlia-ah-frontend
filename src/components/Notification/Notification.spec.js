import React from 'react';
import { configure, shallow } from 'enzyme';
import { renderHook, act } from '@testing-library/react-hooks';
import axios from 'axios';
import MockAdapter from 'axios-mock-adapter';

import Adapter from 'enzyme-adapter-react-16';
import appConfig from '../../config/appConfig';
import Notification from './Notification';
import NotificationItem from '../NotificationItem/NotificationItem';
import notificationState from '../../hooks/Notification';
import * as notificationApiCalls from '../../api/notification';

const { markAsRead } = notificationApiCalls;

const { BACKEND_PATH } = appConfig;

configure({ adapter: new Adapter() });

describe('Notification component', () => {
  let wrapper;
  const notifications = [{
    actor: 'Benjamin',
    message: 'hey',
    novelTitle: 'hello',
    novelUrl: 'test',
    createdAt: '2019-09-09T15:33:29.251Z',
    isRead: false,
  }];
  const mockCallBack = jest.fn();
  beforeEach(() => {
    wrapper = shallow(
      <Notification markAllAsRead={mockCallBack} notifications={notifications} />,
    );
  });
  it('renders correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });
  it('renders should have a span with text Notification', () => {
    expect(wrapper.find('span').text()).toContain('Notification');
  });
  it('renders only one notification item', () => {
    expect(wrapper.find(NotificationItem)).toHaveLength(1);
  });
  it('renders only one notification item', () => {
    wrapper = shallow(
      <Notification markAllAsRead={mockCallBack} notifications={[]} />,
    );
    expect(wrapper.find(NotificationItem)).toHaveLength(0);
  });
  it('call mark all as read once', () => {
    wrapper.find('button').simulate('click');
    expect(mockCallBack.mock.calls.length).toEqual(1);
  });
});
describe('TEST FOR Notification HOOKS', () => {
  it('should update email', () => {
    const { result } = renderHook(() => notificationState([]));
    act(() => {
      result.current.setNotification([{
        actor: 'Benjamin',
        message: 'hey',
        novelTitle: 'hello',
        novelUrl: 'test',
        createdAt: '2019-09-09T15:33:29.251Z',
        isRead: false,
      }]);
    });

    expect(result.current.notifications).toHaveLength(1);
  });
});

describe('TEST API CALLS RESPONSE', () => {
  it('update state after read', () => {
    const { result } = renderHook(() => notificationState([]));
    act(() => {
      result.current.setNotification([{
        actor: 'Benjamin',
        message: 'hey',
        novelTitle: 'hello',
        novelUrl: 'test',
        createdAt: '2019-09-09T15:33:29.251Z',
        isRead: false,
      }]);
    });
    const mock = new MockAdapter(axios);
    const url = `${BACKEND_PATH}/notification`;
    mock.onAny(url).reply(() => {
      act(() => {
        result.current.setNotification([{
          actor: 'Benjamin',
          message: 'hey',
          novelTitle: 'hello',
          novelUrl: 'test',
          createdAt: '2019-09-09T15:33:29.251Z',
          isRead: true,
        }]);
      });
    });
    markAsRead(0, result.current.notifications, result.current.setNotification);

    expect(result.current.notifications[0].isRead).toBe(false);
  });
});
