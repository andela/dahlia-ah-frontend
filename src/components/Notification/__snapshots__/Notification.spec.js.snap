// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Notification component renders correctly 1`] = `ShallowWrapper {}`;
